### DISCUSSION TITLE
Added: LDM Notes for Jan. 15, 2020 and Jan. 22, 2020

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#3146] [General]
Jan 22, 2020

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-01-22.md

1.    Top-level statements and functions
1.    Expression Blocks

Jan 15, 2020

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-01-15.md

Records

 1.   "programming with data"
 1.   Decomposing subfeatures of records

**Comment by orthoxerox** [comment]
Were nondestructive mutation and data-friendly defaults not discussed during the meeting or are the notes from Jan 15 incomplete?

**Comment by qrli** [comment]
> The most notable restrictions are that you cannot branch out, meaning that return, yield break, and yield return, and goto would be illegal in this proposal.

Does it mean `break;` is legal? E.g.
```cs
while (true) {
    Foo({ break; 2 });
}
```

If that's true, we can have something like below?
```cs
while (true) {
    Foo({ break 2; });
    Foo({ break; 2 });
}
```

**Comment by qrli** [comment]
> Object initializers
> ...
> On the other hand, this also removes one of the main benefits for having a constructor, namely that you can validate the whole state of the object before producing it.

Not if object initializer calls an optional method (e.g. `void OnConstructed()`) to do what can be done in a constructor. Similar to what deserializers do after deserialization, that the `[OnDeserialized]` attribute marks a method to be the callback. Maybe it is worth to have that pattern supported by the language itself?

**Comment by agocke** [comment]
@orthoxerox They weren't discussed until later, see https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-01-29.md
